### ISSUE TITLE
Better integration of color corecting software into DT

### ISSUE THREAD

**Issue by matrixik** [darktable-org/darktable#5618]
It's mostly taken from discuss topic so that it won't completely disappear: https://discuss.pixls.us/t/colorchecker-auto-color-luminosity-correction-is-this-currently-possible/17423

**Is your feature request related to a problem? Please describe.**

Now to create new style you need to go between darktable and darktable-chart when using something like Colorchecker, Cmp Digital Target or anything similar.
https://pixls.us/articles/profiling-a-camera-with-darktable-chart/
It have many steps and is slow. At the end we will have many unnecessary profiles that would be used only once.

**Describe the solution you'd like**

1. Select photos (or even only one photo)
2. Trigger option "Profile colors for selected photos"
3. Select photo with color tool (like Colorchecker)
    1. Selecting image could open automatically new window or then could be one more click to trigger it
4. Put grid marks on proper colors (there could be drop down list for known color checker tools, like shown here https://youtu.be/mZvrj8__5r0?t=371 that will show grid with proper number of lines/rows, always open with last used as selected)
    1. Would be even better if image recognition could find it by itself
5. Click "Generate and apply"
6. Done (all images have new color profile applied)

At this point this color profile is useless for any other image so I would prefer if it's not added to any list of styles or similar. It could be stored only with selected images.

**Alternatives**

Not sure, maybe it could also be incorporated into tethering if work on it start again #2930
Probably mostly useful for product and beauty shots.
Take tethered shot with Colorchecker and immediately create color profile from it that will be applied to all following it tethered shots.

**Additional context**

My comment from https://discuss.pixls.us/t/colorchecker-auto-color-luminosity-correction-is-this-currently-possible/17423/16?u=matrixik

> There are quite popular siblings in Poland that are doing weeding photography. In they course about colors ([“Władca barw”](https://sklep.dr5000.com/produkt-MOV0016.html) cost around 10€ but only in Polish) they show how they are using Colorchecker.
> They are photographing it many times per one weeding. Inside church in few places (could have many different light lamps), in the evening (when street lamps have different kind of light sources), in the hall.
> ![Screenshot from 2020-04-10 11-43-54](https://user-images.githubusercontent.com/201930/85406186-b61c8900-b561-11ea-8f3d-2e8ebe74ceae.png)
> They use global profiles (dual illuminant, with custom changes they would do inside Lightroom almost every time, in they words “profile from profile, changing proper colors to nice looking colors”) to make photos from Nikon, Olympus, Fuji and Sony looking the same and then using individual profiles for group of photos from each event (or even only for one photo). Thanks to this most of they photos they simply mass import to Lightroom and then export to jpeg without doing anything. If there is need then they use tool from x-rite to generate profile from Colorchecker and apply it only to few photos that need it, standard adjustments and done. Then they could forget about this profile.
> So they can change this picture
> ![Screenshot from 2020-04-10 12-43-14](https://user-images.githubusercontent.com/201930/85406161-aef57b00-b561-11ea-9867-82780aec6296.png)
> into this with few clicks
> ![Screenshot from 2020-04-10 12-43-21](https://user-images.githubusercontent.com/201930/85406145-a7ce6d00-b561-11ea-8cbb-395d922722b3.png)
> 
> This looks like really fast workflow to me. They are providing they clients with hundreds ready, good looking photos in few days after weeding. Would be nice to have option like “Generate profile/style for individual picture(s)” in dt.
>

**Comment by matrixik** [comment]
Relevant comment from [Jonas_Wagner](https://discuss.pixls.us/t/colorchecker-auto-color-luminosity-correction-is-this-currently-possible/17423/26?u=matrixik)

> Having a quick workflow for dealing with color charts would be awesome. Makes me wonder how much it would take to ‘fully automate the process’, possibly with a bit of cheating and sticking AR markers on the target.
>
> Select a few images in the dark room, including one that contains a color chart, select ‘auto calibrate’ which looks through all the images for the marker (or the color checker directly), creates a profile, applies it to all the selected images. Quick scene referred images even under difficult illumination.
> 
> Should make for a fun project for someone stuck at home.

**Comment by hatsunearu** [comment]
Do you have any sample images with the color checker in-view?

**Comment by matrixik** [comment]
I don't have Colorchecker but I have [CMP Digital Target Mini](http://www.cmp-color.fr/eng%20digital%20target.html)

Quick samples with natural light (rainy today), LED light and salt lamp :P JPG from camera also included (not sure if needed):
https://u.pcloud.link/publink/show?code=XZKHiUkZGEUhIqNihjkxiEaewXMDuuOXiLW7

**Comment by aurelienpierre** [comment]
Generating a LUT from a 49 patches color checker is brittle and generally doesn't work. LUT color profiles are more accurate when produced properly but backfire harder when they are not.

What do you need color correction for ? White balancing ? Producing a color matrice ?

**Comment by matrixik** [comment]
I wrote it in the first post: to get proper colors on photos that was taken in bad conditions, fast.

Also when you check product photography tutorials many of them first thing is using Colorchecker and suggest to use it every day. You would like to "sell" darktable to professionals, and most professionals I talked to use Colorchecker all the time to get every jewelry, dress, car or anything else to have exactly same color on photos like in reality under any lights (and then do some cheating like instead of taking 5 photos of one chair in all 5 colors just make one photo, fix color, clean like remove all dust, threads, scratches etc. then create 4 more copies with recoloring using real life fabric color samples that also need to be in proper color).

**Comment by matrixik** [comment]
> Generating a LUT from a 49 patches color checker is brittle and generally doesn't work. LUT color profiles are more accurate when produced properly but backfire harder when they are not.

I understand that you are passionate about correct way of doing things, and as a programmer myself I think that's great but most people are not like that. Don't take it as an attack on you.

Check x-rite site description of ColorChecker Passport Photo 2 and how they are describing it:
From: https://xritephoto.com/ph_product_overview.aspx?id=2572&catid=158

>Every journey to Color Perfection starts with a Passport!
> 
> The new and more powerful ColorChecker Passport Photo 2 is the indispensable tool that helps you capture the colors of the world more perfectly and ensures that from initial capture to edit, you are beginning with the most accurate files possible. No matter where you set off to explore and create, always have your Passport.
> 
> ColorChecker Passport Photo 2 is ideal for any photographer looking for accurate, consistent color and creative flexibility. The combination of targets and camera calibration software combined with your choice of Lightroom®, Photoshop®, CaptureOne® or Canon® DPP, delivers the most professional quality color control possible.
> 
> It’s a passport to a more perfect place, to a stress-free and consistent state, to a land of infinite creative possibilities.
> 
> ColorChecker Passport Photo 2 is an indispensable tool that helps you capture the colors of the world more perfectly and ensures that from capture to edit, you are beginning with the most accurate files possible.
> 
> The art of color management is all about getting your colors to match from input to output. That means your camera captures true colors, your monitor displays them accurately, and your printer produces a photo that matches what you see on screen.

And most professional photographers believe this literally and they customers also believe it literally and force photographers to use it. From my (outsider) point of view using ColorChecker is standard there and it's not nice to have addition.

**Comment by matrixik** [comment]
On more thing about using ColorChecker (last one, I promise to not spam this issue any more).

In Poland on 5 workshops and one conference I heard from instructors: 
"Replace your ColorChecker when the paint on it stop stinking, colors change because of paint oxidation". 
Is that true? I have no idea, actually it doesn't looks like any of them know that for sure. But they believe in that and replace them.
Apparently X-rite could also check your ColorChecker and give it another certificate if it's still color correct.

**Comment by Egocentrix** [comment]
My first thought is that it is probably possible to do this with a Lua script. Use it to call darktable-chart and apply the created LUT to the selected images.
Instead of a LUT it is also possible to create an icc profile from your image with dcamprof. I'm not an expert, but I think a lot of that can be scripted. An issue with that is that currently darktable needs a restart to update its colour profiles.

**Comment by rabauke** [comment]
I tried to create a custom color profile for my camera using the mentioned colorchecker. As a result of my attempts to create an ICC profile I came to the conclusion that this feature request would be rather challenging to implement. Obviously, the darktable developers cannot come up a new camera profiling tool from scratch. They would have to rely on some existing open source tools, e.g. scanin and colprof of the Argyll suite.

Working with scanin and colprof is pretty cumbersome. (I have not jet tried dcamprof, which also needs scanin for preprocessing.)

* scanin is not very robust in detecting the colorchecker within the image. Careful cropping is needed. 
* colprof requires (as far as I know) that one tweaks the exposure (in post) such that the lightens of the gray patches of the colorchecker mach the reference values.
* Argyll has no reference color values for the colorchecker passport 2 included (Argyll provides a reference file for the older colorchecker passport. Apparently, reference values have been obtained by measuring one speciem, which has been produced in 2012 or before. X-Rite has changed the reference colors in 2014 for regulatory and compliance reasons.).
* I was able to create several profiles (using different options for colprof). I can observe a differencne in the color reproduction using either darktable's standard matrix or the created custom color profiles. There is, however, only a small improvement in color accuracy, I would say. In particular, darktable's standard matrix as well as the created custom color profile have the similar problems in reproducing some very vivid red/orange/pink tones, which get some magenta cast. 

Finally, I created my ICC color profile using the X-Rite colorchecker software. (This profile has a builtin basecurve which follows approximately a gamma correction with gamma = 1.8. Thus, one has to enable the "unbreak input profile" module with gamma = 1/1.8 = 0.5556 and a small positive linear offset of about 0.025 when applying a ICC profile created via the X-Rite software.) I was not jet able to test this profile extensively in various photographic situations. According my preliminary tests, however, employing this profile results in a notable improvement in color reproduction, in particular for highly saturated color tones. Furthermore, colors in darktable (no color modifying modules enabled) match much better the colors of the out-of-cam jpeg (standard color profile).

In summary, I think scanin and colprof are not robust enough to allow a straight forward implementation of color profiling in darktable, e.g., implemented as a Lua script, which calls scanin and colprof for automatic profiling.

Possibly, profiling is beyond the scope of darktable. Instead, one might consider to enhance darktable by enabling darktable to recognize new ICC profiles in `~/.config/darktable/color/in/` without restarting the software. This would allow 

* to export an image for profiling,
* to generate a profile with any external profiling tool,
* to import the color profile

without disrupting work in darktable too much.

Disclaimer: The company that I work for and X-Rite / Pantone are owned by the same organization. I am not an expert in camera profiling and color management. I just described my (limited) personal experience in camera profiling, I do not intend to promote a particular commercial product.

**Comment by jerryhard** [comment]
Same problem i downgrade to 2.6.3 .... because 3.0x us useless ... all version, include GIT 

Color on NEF not working, is total out and if i move with color temperature, then get all to total hell

**Comment by TurboGit** [comment]
First time I see a troll on this forum, congrats @jerryhard !

I have even a better proposal for you, just leave dt and use another software but please don't write ill-informed comments on this forum. Thanks.

**Comment by aurelienpierre** [comment]
> I understand that you are passionate about correct way of doing things, and as a programmer myself I think that's great but most people are not like that. Don't take it as an attack on you.
> 
> Check x-rite site description of ColorChecker Passport Photo 2 and how they are describing it:

It's not about doing things the right or the hacky way, it's about going from a clearly defined problem to a sound, robust and practical solution. Here I feel the problem is ill-defined since it is only about bending the software to introduce the use of a toy into the process. Forget about color checkers. Forget about tools. What is the end user goal in terms of image result ?

If it is to fix the white balance, then we could solve a least-square regression problem of the color checker samples against theoritical values and for a diagonal matrice of WB coefficients.

If it is to fix the colors in general, then we could solve another least-square regression problem of the color checker WB-corrected samples against theoritical values for a full 3×3 matrice.

But that might push colors completely out of gamut and make things worse. So you might want to use a LUT instead of a 3×3 matrice to get, at least, locally minimized error. But for a LUT, a color checker is not enough patches/samples.

That is the conundrum here. On-the-fly/ad-hoc color profiling is great when it works, and worse than nothing if it doesn't. Trying to make it more accurate makes it also less robust, so it might be better when it's good, and much worse otherwise. How do we know if it works ? What are the metrics that can be used to automatically assess the quality of the profiling ? Or do we give users a ticking bomb and say "if you are lucky, it will not explode yet" ?

**Comment by aurelienpierre** [comment]
> According my preliminary tests, however, employing this profile results in a notable improvement in color reproduction, in particular for highly saturated color tones. Furthermore, colors in darktable (no color modifying modules enabled) match much better the colors of the out-of-cam jpeg (standard color profile).

Improvement being more pleasing colors, or more accurate to reality ?

Remember that colors are 100% linked to perception and don't exist outside of human brain, so we need to be careful about what we think we see. Also, color memory tends to embellish reality or even make us correct wrong colors in conjunction with references we have (https://en.wikipedia.org/wiki/Memory_color_effect).

**Comment by rabauke** [comment]
> Improvement being more pleasing colors, or more accurate to reality ?

I meant more accurate to reality (and in this way also more pleasing). 

I am aware of these effects of human color perception that you mentioned. I think, however, the color shifts to magenta that I observe in some highly saturated red/orange tones (when using the standard matrix profile) are not related to these effects. I have observed this issue under various light conditions.

**Comment by aurelienpierre** [comment]
So I guess we could use the darktable-chart tool for samples matching, find the best WB coeffs that minimize the error, input them as params for the white balance module, then find the 3×3 matrice that minimize the error after white-balancing, and input it as params for the channel mixer. 

The complete equation will be :
output = input × WB × ICC × CC
where:
* output is a 3×1 RGB vector,
* WB is a diagonal 3×3, solved from samples against reference, applied is white balance module,
* ICC is the usual color profile (Adobe DNG converter matrice) or possibly an identity matrice, applied in input color profile,
* CC is the ad-hoc color correction, solved from white-balanced samples, applied in channel mixer.
* × denotes matrice product.

This way, user gets 2 ways of manually correcting the computed corrections in case they fail. It's better, in my opinion, than producing a blind ICC profile and works exactly the same, algorithmically.

**Comment by baryluk** [comment]
Hi. I would love for the scanin / colprof be integrated into darktable. It is not uncommon to have dozens of reference pictures of color chart when doing shots, and it is hard to manage them by hand (find it, demosaic, apply lens corrections, crop, scanin, colprof, copy the profile, name it somehow meaningfully, then apply to images). I think it can be simplified and speed up A LOT by by providing better UI / UX.

It doesn't need to use scanin / colprof really, as these can be reimplemented natively in darktable instead, for better integration, control, and working on different platforms easier.

Not sure how much can be reused from scanin / colprof, (need to check licenses), but they do pretty good job. There is room for improvements (like better autodetection and locating the charts, possibly using neural networks or some other computer vision methods).

Right now I use ColorChecker Passport Photo 2 (mostly for outdoors), and Datacolor SpyderCheckr Color (mostly in my studio), and scanin, but it is tedious.

Thanks!

**Comment by aurelienpierre** [comment]
![Screenshot_20201130_100225](https://user-images.githubusercontent.com/2779157/100589472-710c3800-32f3-11eb-9e2f-1a1004b139f3.png)
In progress through color calibration module.

**Comment by aurelienpierre** [comment]
> There is room for improvements (like better autodetection and locating the charts, possibly using neural networks or some other computer vision methods).

That's probably a good idea if you want to use your PC as a heater, but really overkill and not necessary. It's just a least-squares solving in a 3D vector space.

**Comment by rabauke** [comment]
@aurelienpierre What kind of reference values are used for the calibration?

**Comment by aurelienpierre** [comment]
The .cht files from Argyll CMS. Same as darktable-chart, actually.
